feat: Implement baggage span tags feature

zacharycmontoya opened the pull request:
## Motivation
There are a set of W3C baggage items that the Datadog APM product, RUM product, etc. use to provide a comprehensive observability experience, which includes `user.id`, `account.id`, and `session.id`. To deliver on the APM experience, we must make sure that the values are copied over to our web server spans as span tags.

We also want to provide users the ability to customize this feature or disable if desired.

## Changes
Adds new functionality to the Datadog NGINX module to automatically copy W3C baggage items as span tags to the `nginx` spans, following this [RFC](https://docs.google.com/document/d/1ZsgAcQHKjaBxDmhHup0zqFz07WYiOrbBksHPoXJe6u4/edit?tab=t.0#heading=h.q69qzmye8dfb).

This is implemented using two new NGINX directives, copied directly from the `doc/API.md` markdown file since the reference documentation should stand on its own:

---
### `datadog_baggage_tags_enabled`
- **syntax** `datadog_baggage_tags_enabled on|off`
- **context**: `http`, `server`, `location`

If `on`, enable the `datadog_baggage_tags_keys` feature in the current configuration context.
If `off`, disable the `datadog_baggage_tags_keys` feature in the current configuration context.

### `datadog_baggage_tags_keys`
- **syntax** `datadog_baggage_tags_keys all | select <key> [<key> ...]`
- **default** `select user.id account.id session.id`
- **context**: `http`, `server`, `location`

Set span tags for items in the current W3C Baggage header.

To set span tags for baggage items with a matching `<key>`, use `datadog_baggage_tags_keys select <key> [<key> ...]`.

To set span tags for all baggage items, use `datadog_baggage_tags_keys all`.

zacharycmontoya commented on src/datadog_conf.h:
```diff
@@ -215,6 +217,8 @@ struct datadog_loc_conf_t {
   ngx_http_complex_value_t *service_version = DD_NGX_CONF_COMPLEX_UNSET;
   std::unordered_map<std::string, ngx_http_complex_value_t *> tags;
+  ngx_flag_t baggage_span_tags_enabled = NGX_CONF_UNSET;
+  std::unordered_set<std::string> baggage_span_tags;
```
Offline feedback: Use vector instead of unordered_set

zacharycmontoya replied:
Done in e07bfa5549a0587d748d037bd851c66ea93a1198

zacharycmontoya commented on doc/API.md:
```diff
@@ -179,6 +179,14 @@ whose value is the result of evaluating the specified `<value>` in the context
 of the current request.  `<value>` is a string that may contain
 `$`-[variables][2] (including those provided by this module).
 
+### `datadog_baggage_span_tag`
```
I updated `doc/API.md` with better documentation in 50de5ec7005aca90947a1739746d6dfe379e29b7. If that looks ok, then I think we should keep the existing directive name so it's similar to other Datadog tracing libraries. WDYT @dmehala  ?

dmehala reviewed (commented):
I’ve added a few comments, in particular, I believe the wildcard case needs some rework. Additionally, I noticed a couple of scenarios that are not yet covered by tests:

- `datadog_baggage_span_tags user.id;` and `datadog_baggage_span off;` should not set span tags.
- `datadog_baggage_span_tags user.id * fancy.tag;`, ensure that in this case all baggage is captured.

zacharycmontoya commented on src/request_tracing.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <sstream>
 #include <stdexcept>
 #include <string>
+#include <unordered_set>
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on test/cases/baggage_span_tags/test_baggage_span_tags.py:
```diff
@@ -0,0 +1,136 @@
+
+    def test_custom_in_server(self):
+        return self.run_custom_tags_test("./conf/custom_in_server.conf",
+                                         "/http", ["snazzy.tag", "fancy.tag"])
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on test/cases/baggage_span_tags/test_baggage_span_tags.py:
```diff
@@ -0,0 +1,136 @@
+
+    def test_custom_in_http(self):
+        return self.run_custom_tags_test("./conf/custom_in_http.conf", "/http",
+                                         ["snazzy.tag", "fancy.tag"])
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on test/cases/baggage_span_tags/test_baggage_span_tags.py:
```diff
@@ -0,0 +1,136 @@
+            ["user.id", "session.id", "account.id"])
+
+    def test_main_disabled(self):
+        return self.run_custom_tags_test("./conf/disabled.conf", "/http", [])
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on test/cases/baggage_span_tags/test_baggage_span_tags.py:
```diff
@@ -0,0 +1,136 @@
+    def test_overwite_default_tags_empty_inherits_previous_conf(self):
+        return self.run_custom_tags_test(
+            "./conf/builtins.conf", "/http",
+            ["user.id", "session.id", "account.id"])
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on example/tracing/services/nginx/nginx.conf:
```diff
@@ -24,6 +24,9 @@ http {
             # Add a custom tag.  Tag values can contain nginx variables.
             datadog_tag special.tag "The URI is $uri";
+            # Add baggage span tags. If a baggage item has a matching key, the key-value is set a span tag.
+            datadog_baggage_span_tag user.id
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on example/tracing/services/nginx/nginx.conf:
```diff
@@ -33,6 +36,12 @@ http {
+            # Do not configure additional baggage span tags. When unspecified, this is the same as specifiying the following:
+            # datadog_baggage_span_tag user.id
+            # datadog_baggage_span_tag session.id
+            # datadog_baggage_span_tag account.id
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on src/ngx_http_datadog_module.cpp:
```diff
@@ -497,6 +500,17 @@ static char *merge_datadog_loc_conf(ngx_conf_t *cf, void *parent,
+  } else if (conf->baggage_span_tags.empty()) {
+    const auto baggage_span_tags = TracingLibrary::default_baggage_span_tags();
+    for (const auto &tag_name : baggage_span_tags) {
+      conf->baggage_span_tags.emplace_back(std::string(tag_name));
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on src/request_tracing.cpp:
```diff
@@ -74,6 +75,35 @@ static void add_status_tags(const ngx_http_request_t *request, dd::Span &span) {
+    if (conf->baggage_span_tags.size() == 1 &&
+        conf->baggage_span_tags.front() == "*") {
+      baggage.visit([&span](std::string_view key, std::string_view value) {
+        span.set_tag(baggage_prefix + std::string(key), value);
```
Done in 7bec5126aaeeca78c2ab7546f02b8b43a2d75114

zacharycmontoya commented on src/request_tracing.cpp:
```diff
@@ -74,6 +75,35 @@ static void add_status_tags(const ngx_http_request_t *request, dd::Span &span) {
+      baggage.visit([&span](std::string_view key, std::string_view value) {
+        span.set_tag(baggage_prefix + std::string(key), value);
+      });
+    } else {
```
Applied this variant logic in 01316b06d5d18ac8dfc0c501ac12a198314f7618. LMK what you think

zacharycmontoya commented on doc/API.md:
```diff
@@ -179,6 +179,31 @@ whose value is the result of evaluating the specified `<value>` in the context
+If `off`, disables the `datadog_baggage_span_tags` feature in the current configuration context.
+
+This overrides any `baggage_span_tags_enabled` directives at higher levels,
+and may be overriden by `baggage_span_tags_enabled` directives at lower levels.
```
I've removed this in commit 341fdb7835867d32d372a66cb33300180b8e58c0

zacharycmontoya commented on doc/API.md:
```diff
@@ -179,6 +179,31 @@ whose value is the result of evaluating the specified `<value>` in the context
+`baggage.<key>` and whose value is the value of the key-value pair.
+
+This overrides any `baggage_span_tags_enabled` directives at higher levels,
+and may be overriden by `baggage_span_tags_enabled` directives at lower levels.
```
Yes if the configuration is `baggage_span_tags_enabled off` that should disable this. I believe I had a copy/paste issue here, sorry about that!

In commit 341fdb7835867d32d372a66cb33300180b8e58c0 I've reworded this highlighted portion to say the following, which is essentially what you suggested. WDYT?

```
Set span tags for all items from the current W3C Baggage header with a matching `<key>`.
```

zacharycmontoya commented:
> I’ve added a few comments, in particular, I believe the wildcard case needs some rework. Additionally, I noticed a couple of scenarios that are not yet covered by tests:
> 
> * `datadog_baggage_span_tags user.id;` and `datadog_baggage_span off;` should not set span tags.

In b9eee16939390b2db1668abd8a6bd935660149b1 I've added a test case to cover this scenario.

> * `datadog_baggage_span_tags user.id * fancy.tag;`, ensure that in this case all baggage is captured.

For this scenario, I'm getting clarification in the RFC but my reading is that this does not invoke the wildcard behavior. This appears to only be activated when the configuration is exactly `datadog_baggage_span_tags *;` (only one item with the value "*")

zacharycmontoya commented:
> > * `datadog_baggage_span_tags user.id * fancy.tag;`, ensure that in this case all baggage is captured.
> 
> For this scenario, I'm getting clarification in the RFC but my reading is that this does not invoke the wildcard behavior. This appears to only be activated when the configuration is exactly `datadog_baggage_span_tags *;` (only one item with the value "*")

I confirmed with the RFC author that only the configuration `datadog_baggage_span_tags *;` should be a wildcard. If there are multiple items such as `datadog_baggage_span_tags tag1 tag2 *;` then there's no wildcard behavior and the `*` is treated as a named header, since that character is allowed in HTTP headers

zacharycmontoya commented on doc/API.md:
```diff
@@ -179,6 +179,31 @@ whose value is the result of evaluating the specified `<value>` in the context
 of the current request.  `<value>` is a string that may contain
 `$`-[variables][2] (including those provided by this module).
 
+### `datadog_baggage_span_tags`
```
Discussed offline, we decided on `datadog_baggage_tags_enabled` and `datadog_baggage_tags_keys`

zacharycmontoya commented on doc/API.md:
```diff
@@ -179,6 +179,31 @@ whose value is the result of evaluating the specified `<value>` in the context
+As a convenience, you can capture all baggage items as span tags by setting the directive
+`datadog_baggage_span_tags *` rather than specifying each baggage item key.
+
+### `baggage_span_tags_enabled`
```
Discussed offline, this was changed to `datadog_baggage_tags_enabled`

zacharycmontoya commented:
> > then there's no wildcard behavior and the * is treated as a named header, since that character is allowed in HTTP headers
> 
> That means, for the same directive,`*` has two meanings:
> 
> * everything when its by itself.
> * the `*` character when there's multiple items.
> 
> That's error-prone, WDYT?

As discussed offline, to eliminate the `*` character having two meanings, the keys directive must now be specified as `datadog_baggage_tags_keys all` to invoke the wildcard behavior or `datadog_baggage_tags_keys select <key> [<key> ...]` to select keys. So now `*` no longer has any special meanings.

dmehala commented on src/tracing/directives.cpp:
```diff
@@ -63,6 +63,66 @@ char *set_datadog_tag(ngx_conf_t *cf, ngx_command_t *command,
+    return static_cast<char *>(NGX_CONF_ERROR);
+  }
+
+  // If a previous directive usage configured this to use the `all` wildcard,
+  // reset the variant to a vector to hold the new directive's arguments.
```
I would suggest to raise an error instead.

dmehala reviewed (approved):
Except the edge case where a previous call was all and the current one is `select`, the PR LGTM. Thank you.
